### DISCUSSION TITLE
Container crashes / won't start: "No module named 'mailrise'"

### DISCUSSION THREAD

**Discussion by MadKillerChicken** [YoRyan/mailrise#62] [Q&A]
Hi there, and thanks for this great project BTW.

I'm having a strange issue that I've never seen before (not that experienced in the container business).
Container appears to start on startup with the following log output:

```
2023-03-03 18:07:48.899607+00:00 File "/home/mailrise/.local/bin/mailrise", line 5, in <module>
2023-03-03 18:07:48.899628+00:00 from mailrise.skeleton import run
2023-03-03 18:07:48.899638+00:00 ModuleNotFoundError: No module named 'mailrise'
```

Followed the docs, mounted a config file, tried latest and stable variants.
Not sure if I'm missing something here, sounds python, not container related though, any ideas?

**Comment by YoRyan** [comment]
That's really strange. Could I see your full Docker command line?

Are you able to run the container with the bare minimum of arguments?
```
$ docker run -it --rm yoryan/mailrise --help
usage: mailrise [-h] [--version] [-v] [-vv] CONFIG

An SMTP gateway for Apprise notifications

positional arguments:
  CONFIG               path to configuration file

options:
  -h, --help           show this help message and exit
  --version            show program's version number and exit
  -v, --verbose        set loglevel to INFO
  -vv, --very-verbose  set loglevel to DEBUG
```

**Reply by MadKillerChicken** [reply]
I'd love to, thing is, I'm running this on TrueNAS SCALE, which is an UI-driven appliance. I just tell it where the container should be pulled from. Don't think I'd be able to verify what exact command line is used under the hood to start it.

I was hoping you could provide some slight insight into what might be happening code-wise. Sounds like a missing module, which should be unlikely. Could this be caused by some apparently unrelated issue, like, don't know, wrong permission (not running as root) or lack of internet connection (might be trying to pull stuff)?

**Reply by YoRyan** [reply]
I've gotten a few reports from Unraid users about strange Python errors, too. We were eventually able to figure out that Unraid supports _directory_ passthrough but not _file_ passthrough, which is what the image uses by default. So when these users tried to passthrough their /etc/mailrise.conf files, they ended up overriding their entire /etc directory and breaking the Python executable.

I don't entirely understand the error you're getting, but it could be a similar case. Are you willing to share a screenshot of your container configuration?

**Reply by YoRyan** [reply]
So I've just thought of a condition that could cause this error: Are you running the container as the correct user? It should be the non-root user mailrise:mailrise. For enhanced security, the app is installed within the container image as a user-installed Python package. If you were to run the container as, say, root:root, Python would not be able to find the package, and would show the exact error you're getting.

**Reply by MadKillerChicken** [reply]
I think the user issue you're mentioning is most probably the culprit. I actually created a mailrise user but just went for the integrated/default 'app' user. I simply wanted to quickly test the container. Since this usually isn't an issue, as long as permissions are correctly set, I just forgot about it.

I'll try to change user tomorrow and report back. Thank you very much for your input!

**Reply by YoRyan** [reply]
To be clear, users inside containers have no relation to users defined on your host operating system.

The Mailrise image will by default run as the mailrise user it ships with, but TrueNAS may be changing it behind your back, perhaps to root.

**Reply by MadKillerChicken** [reply]
You're obviously right, what I said makes little sense.

Anyway, the only place where user related settings seem to be getting set are called `runAsUser`, `runAsGroup` and `fsGroup` (marked green in the screenshots). All of these are set to the custom user I mentioned earlier, but that's a host OS user. Especially `fsGroup` is used to allow FileSystem access to the config file on the host. The other two settings define the user, that'll be running the pod on the host, I think.

The settings marks in red were disabled, I enabled them to show you all the available security related settings.

I excluded all remaining settings, as they didn't seem relevant (network related and so on).

Please note that setting "runAsNonRoot" doesn't seem to have any effect on this issue.

![01](https://user-images.githubusercontent.com/19254972/223038047-3beaa3ce-4530-4bb7-b7b2-640b2723a892.jpg)

![02](https://user-images.githubusercontent.com/19254972/223038354-16f9a46c-8b8e-4c37-a023-a880ed2dcb25.jpg)

**Reply by abpei** [reply]
> I've gotten a few reports from Unraid users about strange Python errors, too. We were eventually able to figure out that Unraid supports _directory_ passthrough but not _file_ passthrough, which is what the image uses by default. So when these users tried to passthrough their /etc/mailrise.conf files, they ended up overriding their entire /etc directory and breaking the Python executable.
> 
> I don't entirely understand the error you're getting, but it could be a similar case. Are you willing to share a screenshot of your container configuration?

can confirm, had the same issue with portainer. binding the config file only instead of mapping the entire directory as volume fixed it. thanks for this.

**Reply by YoRyan** [reply]
> Anyway, the only place where user related settings seem to be getting set are called `runAsUser`, `runAsGroup` and `fsGroup` (marked green in the screenshots). All of these are set to the custom user I mentioned earlier, but that's a host OS user. Especially `fsGroup` is used to allow FileSystem access to the config file on the host. The other two settings define the user, that'll be running the pod on the host, I think.

Apologies for the long delay—this thread got buried in my notifications. Try `999` for both the user and group. That's the user the container was intended to run as.

**Reply by Hinni79** [reply]
Thank you, thank you, thank you - this mention of 999 finally made the container run on my QNAP NAS. It seems Container Station tries to run as root by default. When using user "mailrise", I wasn't able to access /etc/mailrise.conf. When telling the container to go with user 1000 (my admin user on QNAP), the above python error appeared.

Finally a simple "sudo chown 999:999 mailrise.conf" on the QNAP made the container launch (doesn't matter I have no user with that ID on the NAS). Now I have to figure out, if it works, but at least it does not terminate after a few seconds.

**Reply by Qhilm** [reply]
As a note, GUID '999' is the group 'synopkgs' on DSM 7 (Synology) machines. Not entirely sure how big a security problem that might be.

**Reply by CaptainScully** [reply]
GID 999 is also the group 'openmediavault-config' when running OpenMediaVault, which I would not like to expose to anything else. Maybe support for the `user:` Docker option could let you run the container as a different user/group?
